### PR TITLE
feat: log a message if the previous fetcher failed

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

If the ES fetcher fails and the apm server falls back to kibana, the ES fetcher error is lost.
Add a logger to the chained fetcher and log the error.

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

- run ` TestSourcemapFetcher` systemtest
- check the logs for error messages

### CODE REVIEW

File: internal/sourcemap/chained.go [hunk 1]
```diff
@@ -22,20 +22,35 @@ import (
...
 // will be returned.
-func (c ChainedFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+func (c *ChainedFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
 	var lastErr error
-	for _, f := range c {
+	for _, f := range c.fetchers {
 		consumer, err := f.Fetch(ctx, name, version, path)
+		// log a message if the previous fetcher failed
+		if lastErr != nil {
+			c.logger.With(logp.Error(lastErr)).Debug("failed to fetch sourcemap, falling back to the next fetcher")
```
Comment (axw) [inline internal/sourcemap/chained.go:52]: Would it be worth adding a %T in there to help quickly identify the fetcher type?

**Review by axw — APPROVED** [review]

File: internal/sourcemap/chained.go (hunk 1 above)
Comment (kruskall) [reply internal/sourcemap/chained.go:52]: I'm not sure. The type would always be the sourcemap fetcher. In reality the chained fetcher is only using two fetcher: ES and the fallback to kibana. 
Caching and similar fetchers are just wrapped around them. WDYT ?

Comment (axw) [reply internal/sourcemap/chained.go:52]: Ah OK. Probably not helpful then.

**Review by axw — APPROVED** [review]

**Comment by lahsivjar** [comment]
Tested on branch `8.9` with commit ID `0c34bf69` based on the testing steps in the description. By default, the test doesn't produce any log entries due to how the log entries are handled in systemtest. I updated the test with the below diff and asserted that the log entires are indeed produced:

```diff
diff --git a/systemtest/sourcemap_test.go b/systemtest/sourcemap_test.go
index 06198beff..7eae8ace4 100644
--- a/systemtest/sourcemap_test.go
+++ b/systemtest/sourcemap_test.go
@@ -18,6 +18,7 @@
 package systemtest_test
 
 import (
+	"fmt"
 	"os"
 	"testing"
 
@@ -209,6 +210,18 @@ func TestSourcemapFetcher(t *testing.T) {
 			}
 			err = srv.Start()
 			require.NoError(t, err)
+			defer srv.Close()
+
+			logIter := srv.Logs.Iterator()
+			defer logIter.Close()
+			go func() {
+				for log := range logIter.C() {
+					if log.Logger != "sourcemap" {
+						continue
+					}
+					fmt.Println("SOURCEMAP_LOGS", log.Message)
+				}
+			}()
 
 			// Index an error, applying source mapping and caching the source map in the process.
 			retry := func() {
```

### Output:

```
=== RUN   TestSourcemapFetcher/kibana_fallback_with_es_unreachable
SOURCEMAP_LOGS populating metadata cache
SOURCEMAP_LOGS listening for invalidation...
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
=== RUN   TestSourcemapFetcher/kibana_fallback_with_es_credentials_unauthorized
SOURCEMAP_LOGS populating metadata cache
SOURCEMAP_LOGS listening for invalidation...
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to fetch sourcemap, falling back to the next fetcher
SOURCEMAP_LOGS failed to populate sourcemap metadata: fetcher unavailable: 401 Unauthorized: {"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"API key: invalid ApiKey value","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}}],"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"API key: invalid ApiKey value","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}},"status":401}
```

**Pull request merged**